File: Samples/Digital signatures/VerifySignature/C#/VerifySignature.cs
﻿using System;
using System.Linq;
using System.Text;

namespace BitMiracle.Docotic.Pdf.Samples
{
    public static class VerifySignature
    {
        public static void Main()
        {
            // NOTE:
            // Without a license, the library won't allow you to create or read PDF documents.
            // To get a free time-limited license key, use the form on
            // https://bitmiracle.com/pdf-library/download

            LicenseManager.AddLicenseData("PUT-LICENSE-HERE");

            var sb = new StringBuilder();
            using var pdf = new PdfDocument(@"..\Sample Data\signed.pdf");

            PdfControl? field = pdf.GetControls().FirstOrDefault(c => c.Type == PdfWidgetType.Signature);
            if (field is null)
            {
                Console.WriteLine("Document does not contain signature fields");
                return;
            }

            PdfSignature? signature = ((PdfSignatureField)field).Signature;
            if (signature is null)
            {
                Console.WriteLine("Signature field does not have an associated signature");
                return;
            }

            PdfSignatureContents contents = signature.Contents;
            sb.AppendFormat("Signed part is intact: {0}\n", contents.VerifyDigest());

            DateTime signingTime = signature.SigningTime ?? DateTime.MinValue;
            sb.AppendFormat("Signed on: {0}\n\n", signingTime.ToShortDateString());

            var timestampToken = contents.GetTimestampToken();
            if (timestampToken != null)
            {
                sb.AppendFormat("Is document timestamp: {0}\n", contents.IsDocumentTimestamp);
                sb.AppendFormat("Embedded timestamp: {0}\n", timestampToken.GenerationTime);
                if (timestampToken.TimestampAuthority != null)
                    sb.AppendFormat("Timestamp authority: {0}\n", timestampToken.TimestampAuthority.Name);

                if (!contents.IsDocumentTimestamp)
                    sb.AppendFormat("Timestamp is intact: {0}\n", contents.VerifyTimestamp());
            }

            sb.AppendLine();

            if (contents.CheckHasEmbeddedOcsp())
            {
                sb.AppendLine("Signature has OCSP embedded.");
                CheckRevocation(signature, sb, PdfCertificateRevocationCheckMode.EmbeddedOcsp);
            }

            if (contents.CheckHasEmbeddedCrl())
            {
                sb.AppendLine("Signature has CRL embedded.");
                CheckRevocation(signature, sb, PdfCertificateRevocationCheckMode.EmbeddedCrl);
            }

            CheckRevocation(signature, sb, PdfCertificateRevocationCheckMode.OnlineOcsp);
            CheckRevocation(signature, sb, PdfCertificateRevocationCheckMode.OnlineCrl);

            Console.WriteLine(sb.ToString());
        }

        private static void CheckRevocation(PdfSignature signature, StringBuilder sb, PdfCertificateRevocationCheckMode mode)
        {
            PdfSignatureContents contents = signature.Contents;
            DateTime signingTime = signature.SigningTime ?? DateTime.MinValue;

            foreach (DateTime time in new DateTime[] { signingTime, DateTime.UtcNow })
            {
                bool revoked = contents.CheckIfRevoked(mode, time);
                string status = revoked ? "Revoked" : "Valid";
                string date = time.ToShortDateString();
                sb.AppendFormat("Checking using {0} mode: {1} on {2}\n", mode, status, date);
            }

            sb.AppendLine();
        }
    }
}

File: Samples/Digital signatures/VerifySignature/VB.NET/VerifySignature.vb
Imports System.Text

Imports BitMiracle.Docotic
Imports BitMiracle.Docotic.Pdf

Namespace BitMiracle.Docotic.Pdf.Samples
    Public NotInheritable Class VerifySignature
        Public Shared Sub Main()
            ' NOTE:
            ' Without a license, the library won't allow you to create or read PDF documents.
            ' To get a free time-limited license key, use the form on
            ' https://bitmiracle.com/pdf-library/download

            LicenseManager.AddLicenseData("PUT-LICENSE-HERE")

            Dim sb As New StringBuilder()
            Using pdf As New PdfDocument("..\Sample Data\signed.pdf")
                Dim field As PdfControl = pdf.GetControls().FirstOrDefault(Function(c) c.Type = PdfWidgetType.Signature)
                If field Is Nothing Then
                    Console.WriteLine("Document does not contain signature fields")
                    Return
                End If

                Dim signature = CType(field, PdfSignatureField).Signature
                If signature Is Nothing Then
                    Console.WriteLine("Signature field does not have an associated signature")
                    Return
                End If

                Dim contents = signature.Contents
                sb.AppendFormat("Signed part is intact: {0}" & vbLf, contents.VerifyDigest())

                Dim signingTime = If(signature.SigningTime, Date.MinValue)
                sb.AppendFormat("Signed on: {0}" & vbLf, signingTime.ToShortDateString())

                Dim timestampToken = contents.GetTimestampToken()
                If timestampToken IsNot Nothing Then
                    sb.AppendFormat("Embedded timestamp: {0}" & vbLf, timestampToken.GenerationTime)

                    If timestampToken.TimestampAuthority IsNot Nothing Then
                        sb.AppendFormat("Timestamp authority: {0}" & vbLf, timestampToken.TimestampAuthority.Name)
                    End If

                    sb.AppendFormat("Timestamp is intact: {0}" & vbLf & vbLf, contents.VerifyTimestamp())
                Else
                    sb.AppendLine()
                End If

                If contents.CheckHasEmbeddedOcsp() Then
                    sb.AppendLine("Signature has OCSP embedded.")
                    CheckRevocation(signature, sb, PdfCertificateRevocationCheckMode.EmbeddedOcsp)
                End If

                If contents.CheckHasEmbeddedCrl() Then
                    sb.AppendLine("Signature has CRL embedded.")
                    CheckRevocation(signature, sb, PdfCertificateRevocationCheckMode.EmbeddedCrl)
                End If

                CheckRevocation(signature, sb, PdfCertificateRevocationCheckMode.OnlineOcsp)
                CheckRevocation(signature, sb, PdfCertificateRevocationCheckMode.OnlineCrl)
            End Using

            Console.WriteLine(sb.ToString())
        End Sub

        Private Shared Sub CheckRevocation(signature As PdfSignature, sb As StringBuilder, mode As PdfCertificateRevocationCheckMode)
            Dim contents = signature.Contents
            Dim signingTime = If(signature.SigningTime, Date.MinValue)

            For Each time In {signingTime, Date.UtcNow}
                Dim revoked = contents.CheckIfRevoked(mode, time)
                Dim status = If(revoked, "Revoked", "Valid")
                Dim [date] As String = time.ToShortDateString()
                sb.AppendFormat("Checking using {0} mode: {1} on {2}" & vbLf, mode, status, [date])
            Next

            sb.AppendLine()
        End Sub
    End Class
End Namespace